libraries/Microsoft.Bot.Builder.AI.LUIS/LuisV3Recognizer.cs

```diff
@@ -0,0 +1,510 @@
+
+            BotAssert.ContextNotNull(turnContext);
+
+            if (turnContext.Activity.Type != ActivityTypes.Message)
+            {
+                return null;
+            }
+
+            var utterance = turnContext.Activity?.AsMessageActivity()?.Text;
+            RecognizerResult recognizerResult;
+            JObject luisResponse = null;
+
+            if (string.IsNullOrWhiteSpace(utterance))
+            {
+                recognizerResult = new RecognizerResult
+                {
+                    Text = utterance,
+                    Intents = new Dictionary<string, IntentScore>() { { string.Empty, new IntentScore() { Score = 1.0 } } },
+                    Entities = new JObject(),
+                };
+            }
+            else
+            {
+                var uri = new UriBuilder(_application.Endpoint);
+                uri.Path += $"luis/v3.0-preview/apps/{_application.ApplicationId}";
+
+                var query = AddParam(null, "verbose", options.IncludeInstanceData);
+                query = AddParam(query, "log", options.Log);
+                query = AddParam(query, "show-all-intents", options.IncludeAllIntents);
+                uri.Query = query;
+
+                var content = new JObject();
+                content.Add("query", utterance);
+                if (options.DatetimeReference.HasValue || options.PreferExternalEntities.HasValue)
+                {
+                    var queryOptions = new JObject();
+                    if (options.DatetimeReference.HasValue)
+                    {
+                        queryOptions.Add("datetimeReference", options.DatetimeReference.Value.ToString("o"));
+                    }
+
+                    if (options.PreferExternalEntities.HasValue)
+                    {
+                        queryOptions.Add("overridePredictions", options.PreferExternalEntities.Value);
+                    }
+
+                    content.Add("options", queryOptions);
+                }
+
+                var settings = new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore };
+                if (options.DynamicLists != null)
+                {
+                    foreach (var list in options.DynamicLists)
+                    {
+                        list.Validate();
+                    }
+
+                    content.Add("dynamicLists", (JArray)JsonConvert.DeserializeObject(JsonConvert.SerializeObject(options.DynamicLists, settings)));
+                }
+
+                if (options.ExternalEntities != null)
+                {
+                    foreach (var entity in options.ExternalEntities)
+                    {
+                        entity.Validate();
+                    }
+
+                    content.Add("externalEntities", (JArray)JsonConvert.DeserializeObject(JsonConvert.SerializeObject(options.ExternalEntities, settings)));
+                }
+
+                if (options.Version == null)
+                {
```

Comment:
I don't understand the relationship between options.Version, operation.Slot, and _options.Version and /versions/ and /slots. 

Can you elaborate in a comment here?

Reply:
Done.